vortex-ffi/cinclude/vortex.h

```diff
@@ -0,0 +1,285 @@
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct DType DType;
+
+
+#include <stdarg.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+
+#define DTYPE_NULL 0
+
+#define DTYPE_BOOL 1
+
+#define DTYPE_PRIMITIVE_U8 2
+
+#define DTYPE_PRIMITIVE_U16 3
+
+#define DTYPE_PRIMITIVE_U32 4
+
+#define DTYPE_PRIMITIVE_U64 5
+
+#define DTYPE_PRIMITIVE_I8 6
+
+#define DTYPE_PRIMITIVE_I16 7
+
+#define DTYPE_PRIMITIVE_I32 8
+
+#define DTYPE_PRIMITIVE_I64 9
+
+#define DTYPE_PRIMITIVE_F16 10
+
+#define DTYPE_PRIMITIVE_F32 11
+
+#define DTYPE_PRIMITIVE_F64 12
+
+#define DTYPE_UTF8 13
+
+#define DTYPE_BINARY 14
+
+#define DTYPE_STRUCT 15
+
+#define DTYPE_LIST 16
+
+#define DTYPE_EXTENSION 17
+
+#define LOG_LEVEL_OFF 0
+
+#define LOG_LEVEL_ERROR 1
+
+#define LOG_LEVEL_WARN 2
+
+#define LOG_LEVEL_INFO 3
+
+#define LOG_LEVEL_DEBUG 4
+
+#define LOG_LEVEL_TRACE 5
+
+/**
+ * The FFI interface for an [`Array`].
+ *
+ * Because dyn Trait pointers cannot be shared across FFI, we create a new struct to hold
+ * the wide pointer. The C FFI only seems a pointer to this structure, and can pass it into
+ * one of the various `FFIArray_*` functions.
+ */
+typedef struct Array Array;
+
+/**
+ * FFI-exposed stream interface.
+ */
+typedef struct ArrayStream ArrayStream;
+
+typedef struct File File;
+
+/**
+ * Options supplied for opening a file.
+ */
+typedef struct FileOpenOptions {
+  /**
+   * URI for opening the file.
+   */
+  const char *uri;
+  /**
+   * Additional configuration for the file source (e.g. "s3.accessKey").
+   * This may be null, in which case it is treated as empty.
+   */
+  const char *const *property_keys;
+  /**
+   * Additional configuration values for the file source (e.g. S3 credentials).
+   */
+  const char *const *property_vals;
+  /**
+   * Number of properties in `property_keys` and `property_vals`.
+   */
+  int property_len;
+} FileOpenOptions;
+
+/**
+ * Scan options provided by an FFI client calling the `File_scan` function.
+ */
+typedef struct FileScanOptions {
```

Comment:
this will also hold things like predicate (using the new vortex-expr protobuf) and/or row mask